event-sauce/src/lib.rs

```diff
@@ -182,6 +182,36 @@ where
     }
 }
 
+/// A wrapper trait around [`AggregateUpdate`] to handle event-sauce integration boilerplate
+pub trait UpdateSyncEntityBuilder<ED, EDA>: UpdateEntityBuilder<ED>
+where
+    ED: ConflictCheck<EDA>,
+    EDA: EventData,
+{
+    /// Update the entity with an event
+    fn try_sync_update<B, I>(self, event_builder: B, applied_events: I) -> Result<StorageBuilder<Self, ED>, Self::Error>
+    where
+        B: Into<UpdateEventBuilder<ED>> + Entity,
+        I: std::iter::Iterator<Item=Event<EDA>>,
+    {
+        let event_builder = event_builder.into();
+
+        let event_data = event_builder.get_payload();
+        for ae in applied_events {
+           if let Err(conflict_event_data) = event_data.conflicts_with(ae) {
+               let conflict_event_builder = conflict_event_data.into_builder();
+               return self.try_update(conflict_event_builder);
```

Comment:
XXX: This will find only the first conflicting events, but there potentially may be many more of them, which we ignore (and will process one by one, only as they are resolved?).

Reply:
This may be better as a trait that's implemented on entities. Something like this:

```rust
trait MergeHistories {
    type Error;

    fn merge_histories<AE, IE>(&self, applied_events: AE, incoming_events: IE) -> Result<Self, Self::Error> 
        where 
            AE: IntoIterator<...>,
            IE: IntoIterator<...>;
}
```

Each entity can then define its own conflict resolution behaviour (ignore conflicts, fail at first, collect all conflicts in a vec, etc). The boilerplate is reduced by the `conflicts_with` method implemented on the events.

There's some boilerplate in looping through the `M:N` events, but it's not much as far as I can see.

We also need to consider how the `AE` and `IE` lists are fetched. @Istar-Eldritch did you have any ideas? I guess we might need to add some more methods to the storage traits to get a list of events for an entity ID starting from a given time.

Reply:
`IE`: these are the ones read from the input (created from commands sent to the synchronization endpoint)
`AE`: these will need to be fetched from the data store by the caller of the function, who knows the entity ID as well as the time of the last sync, and has the transaction as well (should be straightforward given these assumptions?)